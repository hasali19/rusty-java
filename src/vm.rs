use std::collections::HashMap;
use std::fs::File;
use std::io::{self, BufReader};
use std::iter;
use std::path::Path;

use bumpalo::Bump;
use color_eyre::eyre::{self, eyre, Context};

use crate::call_frame::CallFrame;
use crate::class::{Class, Method};
use crate::class_file::MethodAccessFlags;
use crate::reader::ClassReader;

pub struct Vm<'a> {
    arena: &'a Bump,
    classes: HashMap<&'a str, &'a Class<'a>>,
    pub(crate) stdout: &'a mut dyn io::Write,
    pub(crate) heap: Bump,
}

impl<'a> Vm<'a> {
    pub fn new(arena: &'a Bump, stdout: &'a mut dyn io::Write) -> Vm<'a> {
        Vm {
            arena,
            classes: HashMap::new(),
            stdout,
            heap: Bump::new(),
        }
    }

    pub fn load_class_file(&mut self, name: &str) -> eyre::Result<&'a Class<'a>> {
        let class_name = name.strip_suffix(".class").unwrap_or(name);

        if let Some(class) = self.classes.get(class_name) {
            return Ok(class);
        }

        let path = Path::new(name).with_extension("class");
        let class_file = self.arena.alloc(
            ClassReader::new(self.arena, BufReader::new(File::open(path)?))
                .read_class_file()
                .wrap_err_with(|| eyre!("failed to read class file '{}'", name))?,
        );

        let class = self.arena.alloc(Class::new(self.arena, class_file)?);

        if let Some(clinit) = class.method("<clinit>", "()V")
            && clinit.access_flags.contains(MethodAccessFlags::STATIC)
        {
            self.call_method(class, clinit)?;
        }

        self.classes.insert(class.name(), class);

        Ok(class)
    }

    pub fn call_method(
        &mut self,
        class: &'a Class<'a>,
        method: &'a Method<'a>,
    ) -> eyre::Result<()> {
        CallFrame::new(class, method, iter::empty(), self)?.execute()?;
        Ok(())
    }
}
