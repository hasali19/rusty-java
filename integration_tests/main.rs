#![feature(exit_status_error)]

use std::fs::{self, File};
use std::path::Path;
use std::process::Command;

use bumpalo::Bump;
use color_eyre::eyre::ContextCompat;
use libtest_mimic::{Arguments, Trial};
use rusty_java::vm::Vm;

fn main() {
    let args = Arguments::from_args();
    let tests_dir = Path::new(file!()).parent().unwrap();

    let tests = fs::read_dir(tests_dir)
        .unwrap()
        .flatten()
        .filter_map(|entry| {
            let path = entry.path();
            let ext = path.extension()?.to_str()?;

            if ext == "java" {
                Some(path.file_stem()?.to_str()?.to_owned())
            } else {
                None
            }
        })
        .map(|n| create_trial(n))
        .collect();

    libtest_mimic::run(&args, tests).exit();
}

fn create_trial(name: String) -> Trial {
    Trial::test(name.clone(), move || {
        let arena = Bump::new();
        let mut stdout = Vec::new();
        let mut vm = Vm::new(&arena, &mut stdout);

        let source_file_path = Path::new(file!())
            .parent()
            .unwrap()
            .join(&name)
            .with_extension("java");

        if !check_stamp(&source_file_path) {
            eprintln!("{source_file_path:?} was modified, recompiling");
            Command::new("javac")
                .arg(&source_file_path)
                .status()?
                .exit_ok()?;
            File::create(source_file_path.with_extension("stamp"))?;
        }

        let class_file_path = source_file_path.with_extension("class");
        let class = vm.load_class_file(class_file_path.to_str().unwrap())?;

        vm.call_method(
            &class,
            class
                .method("main", "([Ljava/lang/String;)V")
                .wrap_err("main method not found")?,
        )?;

        let stdout = String::from_utf8(stdout)?;

        insta::assert_snapshot!(name, stdout);

        Ok(())
    })
}

fn check_stamp(path: impl AsRef<Path>) -> bool {
    let path = path.as_ref();
    let stamp_path = path.with_extension("stamp");

    if !stamp_path.exists() {
        return false;
    }

    let mtime = path.metadata().unwrap().modified().unwrap();
    let stamp_mtime = stamp_path.metadata().unwrap().modified().unwrap();

    stamp_mtime > mtime
}
