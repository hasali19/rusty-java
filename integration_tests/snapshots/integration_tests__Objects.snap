---
source: integration_tests/main.rs
expression: stdout
---
integration_tests/Objects$ChildClass {x: 123, y: 1, z: [1, 2, 3], u: hello}
integration_tests/Objects$ChildClass {x: 124, y: 1, z: [1, 2, 3], u: hello}
integration_tests/Objects$ChildClass {x: 124, y: 0, z: [1, 2, 3], u: goodbye}
